testctrl/svc/orch/executor_test.go

```diff
@@ -0,0 +1,221 @@
+	}
+
+	expectedNames := []string{driver.Name, server.Name, client.Name}
+	for _, en := range expectedNames {
+		found := false
+
+		for _, pod := range pods {
+			if strings.Compare(pod.ObjectMeta.Name, en) == 0 {
+				found = true
+			}
+		}
+
+		if !found {
+			t.Errorf("provision did not create pod for component %v", en)
+		}
+	}
+
+	// test provision with an unsuccessful pod
+	fakePodInf = newFakePodInterface(t)
+
+	driver = types.NewComponent(testContainerImage, types.DriverComponent)
+	server = types.NewComponent(testContainerImage, types.ServerComponent)
+	client = types.NewComponent(testContainerImage, types.ClientComponent)
+
+	components = []*types.Component{server, client, driver}
+	session = types.NewSession(driver, components[:2], nil)
+
+	e = NewExecutor(0, fakePodInf, nil)
+	eventChan = make(chan *PodWatchEvent)
+	e.eventChan = eventChan
+	e.session = session
+
+	go func() {
+		for _, c := range components[:2] {
+			eventChan <- &PodWatchEvent{
+				SessionName:   session.Name,
+				ComponentName: c.Name,
+				Pod:           nil,
+				PodIP:         "127.0.0.1",
+				Health:        Ready,
+				Error:         nil,
+			}
+		}
+
+		eventChan <- &PodWatchEvent{
+			SessionName:   session.Name,
+			ComponentName: components[2].Name,
+			Pod:           nil,
+			PodIP:         "",
+			Health:        Failed,
+			Error:         nil,
+		}
+	}()
+
+	if err := e.provision(fakePodInf); err == nil {
+		t.Errorf("unexpected error with failing pod")
+	}
+}
+
+func TestExecutorMonitor(t *testing.T) {
+	cases := []struct {
+		description string
+		event       *PodWatchEvent
+		errors      bool
+	}{
+		{
+			description: "success event received",
+			event:       &PodWatchEvent{Health: Succeeded},
+			errors:      false,
+		},
+		{
+			description: "failure event received",
+			event:       &PodWatchEvent{Health: Failed},
+			errors:      true,
+		},
+	}
+
+	for _, c := range cases {
+		fakePodInf := newFakePodInterface(t)
+
+		driver := types.NewComponent(testContainerImage, types.DriverComponent)
+		server := types.NewComponent(testContainerImage, types.ServerComponent)
+		client := types.NewComponent(testContainerImage, types.ClientComponent)
+
+		components := []*types.Component{server, client, driver}
+		session := types.NewSession(driver, components[:2], nil)
+
+		e := NewExecutor(0, fakePodInf, nil)
+		eventChan := make(chan *PodWatchEvent)
+		e.eventChan = eventChan
+		e.session = session
+
+		go func() {
+			eventChan <- &PodWatchEvent{
+				SessionName:   session.Name,
+				ComponentName: driver.Name,
+				Pod:           c.event.Pod,
+				PodIP:         c.event.PodIP,
+				Health:        c.event.Health,
+				Error:         c.event.Error,
+			}
+		}()
+
+		err := e.monitor()
+		if err == nil && c.errors {
+			t.Errorf("case '%v' did not return error", c.description)
+		} else if err != nil && !c.errors {
+			t.Errorf("case '%v' unexpectedly returned error '%v'", c.description, err)
+		}
+	}
+}
+
+//func TestExecutorClean(t *testing.T) {
```

Comment:
This function seems impossible to test using the kubernetes fake clientset, because the `DeleteCollection` method used by the Executor's `clean` method does not appear to delete in the fake. Subsequent calls to `List` all pods always return the original list.

`DeleteCollection` is more efficient than `Delete`, because it makes only 1 network request to the kubernetes API. This is why it was chosen. It may be better to forgo this efficiency and move to calling `Delete` on each component's pod.

Reply:
Confirmed: kubernetes/client-go#609